src/Microsoft.AspNetCore.DataProtection.EntityFrameworkCore/EntityFrameworkCoreXmlRepository.cs

```diff
@@ -0,0 +1,100 @@
+    /// An <see cref="IXmlRepository"/> backed by an EntityFrameworkCore datastore.
+    /// </summary>
+    public class EntityFrameworkCoreXmlRepository<TContext> : IXmlRepository
+        where TContext : DbContext, IDataProtectionKeyContext
+    {
+        private readonly Func<TContext> _contextFactory;
+        private readonly ILogger _logger;
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="EntityFrameworkCoreXmlRepository{TContext}"/>.
+        /// </summary>
+        /// <param name="services"></param>
+        /// <param name="loggerFactory">The <see cref="ILoggerFactory"/>.</param>
+        public EntityFrameworkCoreXmlRepository(IServiceProvider services, ILoggerFactory loggerFactory)
+            : this(CreateContextFactory<TContext>(services), loggerFactory)
+        {
+        }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="EntityFrameworkCoreXmlRepository{TContext}"/>.
+        /// </summary>
+        /// <param name="contextFactory">The factory method that creates a context to store instances of <see cref="DataProtectionKey"/></param>
+        /// <param name="loggerFactory">The <see cref="ILoggerFactory"/>.</param>
+        public EntityFrameworkCoreXmlRepository(Func<TContext> contextFactory, ILoggerFactory loggerFactory)
+        {
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _contextFactory = contextFactory ?? throw new ArgumentNullException(nameof(contextFactory));
+            _logger = loggerFactory.CreateLogger<EntityFrameworkCoreXmlRepository<TContext>>();
+        }
+
+        /// <inheritdoc />
+        public virtual IReadOnlyCollection<XElement> GetAllElements()
+        {
+            using (var context = _contextFactory())
+            {
+                return context.Set<DataProtectionKey>().AsNoTracking().Select(key => TryParseKeyXml(key.Xml)).ToList().AsReadOnly();
+            }
+        }
+
+        /// <inheritdoc />
+        public void StoreElement(XElement element, string friendlyName)
+        {
+            using (var context = _contextFactory())
+            {
+                var newKey = new DataProtectionKey()
+                {
+                    FriendlyName = friendlyName,
+                    Xml = element.ToString(SaveOptions.DisableFormatting)
+                };
+
+                context.Set<DataProtectionKey>().Add(newKey);
+                _logger.LogSavingKeyToDbContext(friendlyName, typeof(TContext).Name);
+                context.SaveChanges();
+            }
+        }
+
+        private static Func<T> CreateContextFactory<T>(IServiceProvider services)
+        {
+            return () =>
+            {
+                using (var scoped = services.GetRequiredService<IServiceScopeFactory>().CreateScope())
+                {
+                    return scoped.ServiceProvider.GetRequiredService<T>();
```

Comment:
Wouldn't this return the DbContext instance and then immediately after dispose it by disposing the scope that contains it? Perhaps I am missing something or forgetting something about how DI works.